Contract court acting on confirmed chain events

halseth opened the pull request:
This PR changes the way the `ChannelArbitrator` state machine is triggered, by making the `ChainEvents` sent from the `ChainWatcher` only being sent if a commitment is confirmed on-chain. This is done to ensure that if we act on a commitment (force close, breach, cooperative close etc), then the commitment we act on will not be replaced by another, valid commitment in-chain.

### ChannelArbitrator
The `ChannelArbitrator`'s state machine is changed to accept any commitment being seen on-chain, regardless what state it is currently in. In particular, we would previously assume a local force close transaction always would be confirmed if we had successfully broadcasted it. Now we broadcast, and wait for any commitment to be accepted into a block before making any more assumptions.

### OpenChannel
By having to wait for a commitment to appear in-chain before taking any further action, a new "pending close state" a channel can be in is introduced. We add a field `BroadcastedCommitment` to the `OpenChannel` struct, which keeps track of the intermediate state a channel can be in if the commitment has been broadcasted. This field is needed to make sure we don't try to forward payments over such a channel, as it might be our commitment that is broadcasted.

### ChainWatcher
The `ChainWatcher` now only dispatches confirmed events, making all it's clients unaware of any unconfirmed spends.

Fixes #538 
Fixes #880
Fixes #1112 
Fixes #1071 
Fixes #1091

**Note:** builds on top of #956 and #969, and the integration tests need https://github.com/Roasbeef/btcd/pull/20 to avoid flakes.

TODO before merge:
- [x] Finish unit tests for the ChannelArbitrator
- [x] ~~Database migration logic~~
- [x] Update dependencies if/when https://github.com/Roasbeef/btcd/pull/20 is merged

Roasbeef commented:
One integration tests hit what looks to be a deadlock. 


The other instance ran into this error (which may or may not be new?): 
```

    --- FAIL: TestLightningNetworkDaemon/test_multi-hop_htlc_local_chain_claim (21.98s)
    	lnd_test.go:69: Failed: (test multi-hop htlc local chain claim): exited with error: 
    		*errors.errorString transactions not found in mempool: wanted 3, found 2 txs in mempool
```

(or possibly is an incorrect test that was coded around a prior race condition?)

halseth commented:
The "transactions not found in mempool" error should be fixed when rebased on `btcd` with the mempool fix. It is basically an existing race condition that my changes made more likely to occur.

The deadlock I'm not sure about, not seen before 🤔

Roasbeef commented on rpcserver.go:
```diff
@@ -1434,7 +1434,8 @@ func (r *rpcServer) PendingChannels(ctx context.Context,
 		// the utxoNursery for additional information.
-		case channeldb.ForceClose:
+		// TODO(halseth): distinguish remote and local case?
+		case channeldb.LocalForceClose, channeldb.RemoteForceClose:
```
Could do so by extending the RPC layer, possibly left to a follow up PR driven by a new issue.

Roasbeef commented on rpcserver.go:
```diff
@@ -1499,6 +1500,60 @@ func (r *rpcServer) PendingChannels(ctx context.Context,
+			LocalBalance:  int64(pendingClose.LocalCommitment.LocalBalance),
+		}
+
+		switch pendingClose.BroadcastedCommitment {
```
Seems we also need to modify `ListChannels` to _not_ list any channels that have had their commitment transactions broadcast.

halseth replied:
Done.

Roasbeef commented on lnd_test.go:
```diff
@@ -1048,8 +1048,17 @@ func testDisconnectingTargetPeer(net *lntest.NetworkHarness, t *harnessTest) {
-	if err := net.DisconnectNodes(ctxt, net.Alice, net.Bob); err != nil {
-		t.Fatalf("unable to disconnect Bob's peer from Alice's: err %v", err)
+	var predErr error
+	err = lntest.WaitPredicate(func() bool {
```
Why's this necessary now?

halseth replied:
Because we now don't consider the channel closed (adding the `CloseSummary` to the db) before the commitment is on-chain, we run into a race with the above force close. Alternative is to modify the `DisconnectPeer` rpc to allow disconnecting from a peer where the commitment has been broadcasted, but not confirmed yet?

Roasbeef commented on chainntnfs/interface_test.go:
```diff
@@ -460,6 +451,120 @@ func testSpendNotification(miner *rpctest.Harness,
+	}
+
+	// TODO(halseth): create new clients that should be registered after tx
+	// is in the mempool already, when btcd supports notifying on these.
```
`btcd` supports this now.

halseth replied:
Issue created to address this.

Roasbeef commented on channeldb/channel.go:
```diff
@@ -1551,6 +1596,7 @@ type ChannelCloseSummary struct {
 
 	// CloseType details exactly _how_ the channel was closed. Three
 	// closure types are possible: cooperative, force, and breach.
+	// TODO: migration logic.
```
No migration logic needed if we append the new close types to the _end_.

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1348,20 +1357,65 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32,
 
+		// We have broadcasted our commitment, and it is now confirmed
+		// on-chain.
+		case closeSummary := <-c.cfg.ChainEvents.LocalUnilateralClosure:
```
Would need to double check, but I think it's the case that for all of these events, we need to ensure we always pass in the _block height_ of the event. Otherwise, we may end up writing the incorrect height hint to disk, which has ramifications for neutrino (possibly missing an event).

halseth replied:
Now using the `spendingHeight` when advancing state. Also changed the variable name to `triggerHeight` to make it more clear.

Roasbeef commented on contractcourt/channel_arbitrator_test.go:
```diff
@@ -1 +1,407 @@
+	//	}
+}
+
+func TestChannelArbitratorCooperativeClose(t *testing.T) {
```
Reminder to at least finish this pending started set of tests.

halseth replied:
Tests finished.

Roasbeef commented:
Also needs a rebase in light of the recent merges.

halseth commented on channeldb/channel.go:
```diff
@@ -327,6 +327,11 @@ type OpenChannel struct {
+	// BroadcastedCommitment indicates what kind of commitment has been
+	// broadcasted, if any.
+	// TODO: migration logic.
+	BroadcastedCommitment ClosureType
```
Clever! 😀

halseth commented on channeldb/channel.go:
```diff
@@ -595,6 +600,37 @@ func (c *OpenChannel) MarkBorked() error {
+// MarkCommitmentBroadcasted marks the channel as a commitment transaction has
+// been broadcast, either our own or the remote, and we should watch the chain
+// for it to confirm before taking any further action.
+func (c *OpenChannel) MarkCommitmentBroadcasted(closeType ClosureType) error {
```
Was to make it possible to mark breach/remote close if we wanted that in the future. But since we can keep this backwards compatible by creating the `ChanStatus` field, I'll just have one state for now.

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -539,6 +541,27 @@ func (c *ChannelArbitrator) stateStep(bestHeight uint32, bestHash *chainhash.Has
+		// trimmed.  We'll need to wait for a CSV timeout before we can
+		// reclaim the funds.
+		commitRes := contractResolutions.CommitResolution
+		if commitRes != nil && commitRes.MaturityDelay > 0 {
```
Shouldn't the check `commitRes.MaturityDelay > 0` be sufficient?

halseth commented on contractcourt/chain_watcher.go:
```diff
@@ -496,6 +507,34 @@ func (c *chainWatcher) dispatchCooperativeClose(commitSpend *chainntnfs.SpendDet
+	if err != nil {
+		return err
+	}
+	// With the event processed, we'll now notify all subscribers of the
```
Fixed.

halseth replied:
Fixed.

halseth commented on contractcourt/briefcase.go:
```diff
@@ -112,25 +112,31 @@ const (
+	// StateCommitmentBroadcasted is a state that indicates that the
+	// attendant has broadcasted the commitment transaction, and is now
+	// waiting for it to confirm.
+	StateCommitmentBroadcasted ArbitratorState = 2
```
Done.

halseth commented on channeldb/channel.go:
```diff
@@ -1478,23 +1514,32 @@ func (c *OpenChannel) FindPreviousState(updateNum uint64) (*ChannelCommitment, e
+	// None indicates that no closing commitment is broadcasted, or
+	// confirmed.
+	// TODO: not use iota?
+	None ClosureType = iota
```
Done.

halseth commented:
Addressed everything except integration tests and `chanarb` tests.

merehap commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1348,20 +1356,65 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32,
+			// When the state machine reaches that state, we'll log
+			// out the set of resolutions.
+			stateCb := func(nextState ArbitratorState) error {
+				if nextState == StateContractClosed {
```
If you invert this condition, you can save a level of indentation and prevent a lot of line-wrapping.
```go
if nextState != StateContractClosed {
    return nil
}
```

Admittedly you'd have to have two "return nil"s then, but on net it still seems nicer.

halseth replied:
Agreed! 👍

halseth commented on peer.go:
```diff
@@ -329,9 +329,9 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
-		if dbChan.IsBorked {
-			peerLog.Warnf("ChannelPoint(%v) is borked, won't "+
-				"start.", chanPoint)
+		if dbChan.ChanStatus != channeldb.Default {
```
Done

halseth commented on lnd_test.go:
```diff
@@ -1591,21 +1633,12 @@ func testChannelForceClosure(net *lntest.NetworkHarness, t *harnessTest) {
-
-	// The commitment transaction has not been confirmed, so we expect to
-	// see a maturity height and blocks til maturity of 0.
-	assertCommitmentMaturity(t, forceClose, 0, 0)
```
Same for `RecoveredBalance`

halseth commented on contractcourt/chain_watcher.go:
```diff
@@ -251,6 +251,30 @@ func (c *chainWatcher) SubscribeChannelEvents(syncDispatch bool) *ChainEventSubs
 	if syncDispatch {
 		sub.ProcessACK = make(chan error, 1)
+
+		// If this client is syncDispatch, we cannot safely delete it
```
I kept this commit for now, as it is needed to make the integration tests pass reliably. It should be reverted when #1104 is merged.

halseth commented on lnd_test.go:
```diff
@@ -1510,6 +1510,8 @@ func testChannelForceClosure(net *lntest.NetworkHarness, t *harnessTest) {
 		t.Fatalf("unable to execute force channel closure: %v", err)
 	}
 
+	time.Sleep(1 * time.Second)
```
Most sleeps should be eliminated.

halseth commented on channeldb/db.go:
```diff
@@ -377,23 +414,26 @@ func fetchChannels(d *DB, pendingOnly bool) ([]*OpenChannel, error) {
+					if channel.IsPending != pending {
+						continue
+					}
+					if (channel.ChanStatus != Default) != waitingClose {
```
Rewrote and added comments to make it more clear what is going on.

Roasbeef commented on channeldb/channel.go:
```diff
@@ -285,6 +285,38 @@ type ChannelCommitment struct {
+	// Borked indicates that the channel has entered an irreconcilable
+	// state, triggered by a state desynchronization or channel breach.
+	// Channels in this state should never be added to the htlc switch.
+	Borked ChannelStatus = 1
```
That'll be a follow which will lead to a full data-loss protection implementation.

Roasbeef commented:
Needs rebase to resolve proto conflict due to recent merge.

halseth commented:
Rebased and addressed comments, but will attempt to move the `pending close` state from `OpenChannel` to `ClosedChannel` in the database before merge.

Roasbeef commented on channeldb/channel.go:
```diff
@@ -1549,8 +1599,9 @@ type ChannelCloseSummary struct {
-	// closure types are possible: cooperative, force, and breach.
+	// CloseType details exactly _how_ the channel was closed. Four closure
+	// types are possible: cooperative, local force, remote force, and
+	// breach.
```
Was it removed in this diff? Didn't pick up...or are we talking about merging some of the other ones?

halseth replied:
Updated.

halseth commented:
Decided to keep as is. Addressed last comment and rebased 💯
